packages/flutter/lib/src/gestures/mouse_tracking.dart

```diff
@@ -0,0 +1,308 @@
+      : assert(router != null),
+        assert(annotationFinder != null) {
+    router.addGlobalRoute(_handleEvent);
+  }
+
+  /// Used to find annotations at a given logical coordinate.
+  final MouseDetectorAnnotationFinder annotationFinder;
+
+  // The set of annotations that are currently being tracked. They may or may
+  // not be active, depending on the value of _TrackedAnnotation.active.
+  final Set<_TrackedAnnotation> _trackedAnnotations = Set<_TrackedAnnotation>();
+
+  /// Allows the [MouseDetector] to attach an annotation as an active
+  /// annotation.
+  void attachAnnotation(MouseDetectorAnnotation annotation) {
+    _trackedAnnotations.add(_TrackedAnnotation(annotation));
+    // Schedule a check so that we test this new annotation to see if the mouse
+    // is currently inside its region.
+    _scheduleMousePositionCheck();
+  }
+
+  /// Allows the [MouseDetector] to detach an annotation as an active annotation.
+  void detachAnnotation(MouseDetectorAnnotation annotation) {
+    final _TrackedAnnotation tracked = _findAnnotation(annotation);
+    assert(tracked != null, "Tried to detach an annotation that wasn't attached: $annotation");
+    if (tracked.active) {
+      annotation.onExit(MouseExitDetails(
+          globalPosition: lastMouseEvent?.position,
+          sourceTimeStamp: lastMouseEvent?.timeStamp));
+    }
+    _trackedAnnotations.remove(tracked);
+  }
+
+  void _scheduleMousePositionCheck() {
+    SchedulerBinding.instance.addPostFrameCallback((Duration _) => collectMousePositions());
+    SchedulerBinding.instance.scheduleFrame();
+  }
+
+  // Handler for events coming from the PointerRouter.
+  void _handleEvent(PointerEvent event) {
+    if (event.kind != PointerDeviceKind.mouse) {
+      return;
+    }
+    if (_trackedAnnotations.isEmpty) {
+      // If we're not tracking anything, then there is no point in registering a
+      // frame callback or scheduling a frame. By definition there are no active
+      // annotations that need exiting, either.
+      lastMouseEvent = null;
+      return;
+    }
+    if (event is PointerRemovedEvent) {
+      lastMouseEvent = null;
+      // If the mouse was removed, then we need to schedule one more check to
+      // exit any annotations that were active.
+      _scheduleMousePositionCheck();
+    } else {
+      if (event is PointerMoveEvent || event is PointerHoverEvent || event is PointerDownEvent) {
+        if (lastMouseEvent == null || lastMouseEvent.position != event.position) {
+          // Only schedule a frame if we have our first event, or if the
+          // location of the mouse has changed.
+          _scheduleMousePositionCheck();
+        }
+        lastMouseEvent = event;
+      }
+    }
+  }
+
+  _TrackedAnnotation _findAnnotation(MouseDetectorAnnotation annotation) {
+    final _TrackedAnnotation tracked = _trackedAnnotations.firstWhere(
+      (_TrackedAnnotation tracked) => tracked.annotation == annotation,
+      orElse: () => null,
+    );
```

Comment:
can we make it a map instead? you'd have to implement the `difference` logic manually, but it would be more efficient here.